#Перевірка статиcтичних гіпотез. Z-test.



0. Зчитайте дані з `data.csv` у змінну data, яка має тип pandas.DataFrame. Ми будемо далі працювати з цією змінною.

In [5]:
import pandas as pd
import numpy as np
from scipy import stats

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
df_static = pd.read_csv ('/content/drive/MyDrive/Colab Notebooks/Data/data.csv')

Запустіть код нижче. Для коректної робити всіх подальших методів дані мають бути у вигляді одновимірного `numpy.ndarray` вектора та мати розмірність (100,). Така розмірність означає, що дані - одновимірні (колонка), якщо б розмірність була (100, 1), то дані сприймаються методами як двовимірні (таблиця), хоча для нас виглядати можуть так само.

In [4]:
data = df_static.data.values
data.shape

(100,)

**Завдання 1**. Зчитані дані - це сума покупок в доларах наших покупців на сайті протягом липня 2025 року.

До цього місяця, ми мали продажі в середньому на чек $20.

Необхідно зʼясувати, чи вийшло у нас статистично значущо **підвищити** середній чек за липень 2025?

Аби дати відповідь - ми проведемо z-test на рівні значущості $\alpha=0.05$ двома способами. В цьому завданні ми будемо виконувати обчислення "вручну" з використанням бібліотек numpy та scipy.stats подібно до прикладу в лекції. Для цього виконайте наступні 7 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть параметри задачі у змінні Python:
    - `sample_mean` - $\bar{x}$, середнє значення в вибірці
    - `population_mean` - $\mu_0$, середнє значення в популяції (тобто те, з яким ми порівнюємо середнє вибіркове значення)
    - `population_std` - $\sigma$, вибіркове стандартне відхилення, яке ми вважаємо, що є рівним ст. відх. популяції, адже маємо достатньо велику вибірку.
    - `sample_size` - $n$, розмір вибірки
    - `alpha` - $\alpha$ рівень значущості
    
    **Увага!** Для обчислення стандартного відхилення маємо скористатись функцією `np.std(your_dataframe, ddof=1)`. Чому так - розбираємо в лекції про t-test.


2. Визначте для цієї задачі:
    - якою є гіпотеза $H_0$
    - якою є альтернативна гіпотеза $H_a$
    - з яким типом тесту ми маємо справу - лівосторонній, правосторонній чи двосторонній.

3. Обчисліть стандартну помилку SE.
4. Розрахуйте z-статистику.
5. Знайдіть критичне z-значення з допомогою бібліотеки stats.
6. Обчисліть p-value з допомогою бібліотеки stats.
7. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$. Для прийняття рішення зробіть і порівняння z-статистики з критичним значенням, і проаналізуйте p-value.

In [38]:
# Розраховуємо відомі параметри
sample_mean = data.mean()
population_mean = 20
population_std = np.std(data, ddof=1)
sample_size = data.size
alpha = 0.05

In [40]:
#вивід значень
print ('sample_mean = ', sample_mean.round(4))
print ('population_mean = ', population_mean)
print ('population_std = ', population_std.round (4))
print ('sample_size = ', sample_size)
print ('alpha = ', alpha)

sample_mean =  19.3776
population_mean =  20
population_std =  6.2536
sample_size =  100
alpha =  0.05


H0 = 20,

Ha > 20

тест правосторонній

In [41]:
#обчислюємо стандартну помилку
SE = population_std / np.sqrt(sample_size)
print ('SE = ', SE.round(4))

SE =  0.6254


In [42]:
# розраховуємо z - статистику
z_statistic = (sample_mean - population_mean) / SE
print('Z_statistic =', z_statistic.round (4))

Z_statistic = -0.9953


In [43]:
# розраховуємо критичне z - значення
z_critical = stats.norm.ppf(1 - alpha)
print('Z_critical =', z_critical.round (4))

Z_critical = 1.6449


In [44]:
# розраховуємо критичне p - value
p_value = 1 - stats.norm.cdf(z_statistic)
print('p_value =', p_value.round (4))

p_value = 0.8402


In [45]:
print("Порівняння z-статистики з критичним значенням")
print()
print('Z_statistic =', z_statistic.round (4))
print('Z_critical =', z_critical.round (4))
print(f"Порівняння: {z_statistic.round (4)} {'>' if z_statistic > z_critical else '≤'} {z_critical.round (4)}")

if z_statistic > z_critical:
    print("Висновок: z-статистика ПЕРЕВИЩУЄ критичне значення")
    decision1 = "ВІДХИЛЯЄМО H₀"
else:
    print("Висновок: z-статистика НЕ перевищує критичне значення")
    decision1 = "НЕ ВІДХИЛЯЄМО H₀"

print(f"Рішення: {decision1}")

Порівняння z-статистики з критичним значенням

Z_statistic = -0.9953
Z_critical = 1.6449
Порівняння: -0.9953 ≤ 1.6449
Висновок: z-статистика НЕ перевищує критичне значення
Рішення: НЕ ВІДХИЛЯЄМО H₀


In [46]:
print("Порівняння p-value з рівнем значущості")
print()
print('p_value =', p_value.round (4))
print(f"Рівень значущості α = {alpha}")
print(f"Порівняння: {p_value.round(4)} {'<' if p_value < alpha else '≥'} {alpha}")

if p_value < alpha:
    print("Висновок: p-value МЕНШЕ за рівень значущості")
    decision2 = "ВІДХИЛЯЄМО H₀"
else:
    print("Висновок: p-value НЕ менше за рівень значущості")
    decision2 = "НЕ ВІДХИЛЯЄМО H₀"

print(f"Рішення: {decision2}")

Порівняння p-value з рівнем значущості

p_value = 0.8402
Рівень значущості α = 0.05
Порівняння: 0.8402 ≥ 0.05
Висновок: p-value НЕ менше за рівень значущості
Рішення: НЕ ВІДХИЛЯЄМО H₀


H₀ не відхиляємо, так як z - статистика (-0,9953) не перевищує z - критичне (1,645) і p-value (0,8402) не менше alpha =  0.05



**Завдання 2.** Виконайте обчислення z-test з використанням бібліотеки statsmodels.

Отримайте z-статистику та р-значення.

Виведіть p-значення та зробіть висновок, чи ми маємо достатньо доказів, аби стверджувати, що середній чек зріс.

Чи зійшлись значення z-статистику та р-значення в цьому завданні з попередніми обчисленнями?

Завдання оцінюється в 3 бали.

In [33]:
from statsmodels.stats import weightstats as stests


In [49]:
z_stat, p_val = stests.ztest(data,
                            value=population_mean,
                            alternative='larger')

print(f"Результати z-тесту:")
print(f"  Z-статистика: {z_stat.round(4)}")
print(f"  P-value: {p_val.round(4)}")
print()

Результати z-тесту:
  Z-статистика: -0.9953
  P-value: 0.8402



In [50]:
print("Порівняння p-value з рівнем значущості")
print()
print('p_value =', p_val.round (4))
print(f"Рівень значущості α = {alpha}")
print(f"Порівняння: {p_val.round(4)} {'<' if p_val < alpha else '≥'} {alpha}")

if p_val < alpha:
    print("Висновок: p-value МЕНШЕ за рівень значущості")
    decision2 = "ВІДХИЛЯЄМО H₀"
else:
    print("Висновок: p-value НЕ менше за рівень значущості")
    decision2 = "НЕ ВІДХИЛЯЄМО H₀"

print(f"Рішення: {decision2}")

Порівняння p-value з рівнем значущості

p_value = 0.8402
Рівень значущості α = 0.05
Порівняння: 0.8402 ≥ 0.05
Висновок: p-value НЕ менше за рівень значущості
Рішення: НЕ ВІДХИЛЯЄМО H₀


не відхиляємо нульову гіпотезу.

так як p-value (0.8402) суттєво перевищує рівень значущості альфа(0.05), тому у нас немає достатніх статистичних доказів, щоб стверджувати, що середній чек у липні 2025 року зріс.

Дані з першим завданням повністю зійшлися